src/logging.rs

```diff
@@ -97,6 +105,47 @@ pub mod trace {
         }
     }
 
+    impl LevelHandle {
+        /// Returns a new `LevelHandle` without a corresponding filter.
+        ///
+        /// This will do nothing, but is required for admin endpoint tests which
+        /// do not exercise the `proxy-log-level` endpoint.
+        pub fn dangling() -> Self {
+            let builder = subscriber_builder()
+                .with_filter(filter::EnvFilter::default())
+                .with_filter_reloading();
+            let inner = builder.reload_handle();
+            LevelHandle { inner }
+        }
+
+        pub fn set_level(&self, level: impl AsRef<str>) -> Result<(), Error> {
+            let level = level.as_ref();
+            let filter = level.parse::<filter::EnvFilter>()?;
```

Comment:
Where does this parsing code live? It seems like we should definitely have some testing around invalid input data.

Reply:
The parsing code lives [here](https://github.com/tokio-rs/tracing/blob/0fe4d78d9f323079af1022dbc4d53dda87942134/tracing-fmt/src/filter/env.rs#L279-L379); there are [tests](https://github.com/tokio-rs/tracing/blob/0fe4d78d9f323079af1022dbc4d53dda87942134/tracing-fmt/src/filter/env.rs#L526-L804) later on in that file.